### ISSUE TITLE
🐛 BUG: Svelte components using .d.ts declarations with SvelteComponentTyped are not supported

### ISSUE THREAD

**Issue by Princesseuh** [withastro/language-tools#425]
### Describe the Bug

`SvelteComponentTyped`'s constructor expect a `target` attribute to be passed. I am unfortunately not super knowledgeable in Svelte's internals to know what it's used for, but one thing is sure: We don't pass it in Astro files.

Looking at Svelte2tsx tests, it seems like passing it a dummy value is okay, but the problem is that unlike Svelte, we don't know if a components needs it or not (and passing it to all components would lead to errors for other frameworks). 

A possible solution would be to patch the definition of `SvelteComponentTyped` when we load the snapshot for Svelte's `dev.d.ts`, but I wouldn't be surprised if that brings in more issues 🤔

### Steps to Reproduce

1. Import a component from a Svelte library with typings, ex: `import { Theme } from "carbon-components-svelte"`
2. Use it in the template
3. Error! Says `target` is required even though that props doesn't exist

**Comment by pfloride** [comment]
> I'm fighting this issue for the last couple of hours.
> 
> The workaround is not working for me, only the error message is slightly different depending on the used import method.
> 
> Importing with `import {Heading} from 'svelte-example-library'` results in:
> 
> ```
> Type '{ children: string; variant: string; }' is not assignable to type 'IntrinsicAttributes & ComponentConstructorOptions<{ [x: string]: any; variant?: Variant | undefined; class?: string | null | undefined; }>'.
>   Property 'variant' does not exist on type 'IntrinsicAttributes & ComponentConstructorOptions<{ [x: string]: any; variant?: Variant | undefined; class?: string | null | undefined; }>'.ts(2322)
> ```
> 
> Importing with `import Heading from 'svelte-example-library/Heading.svelte'` results in:
> 
> ```
> Type '{ children: string; variant: string; }' is not assignable to type 'IntrinsicAttributes & ComponentConstructorOptions<Record<string, any>>'.
>   Property 'variant' does not exist on type 'IntrinsicAttributes & ComponentConstructorOptions<Record<string, any>>'.ts(2322)
> ```
> 
> In both cases, the component is declared like this `<Heading variant="title">Some Title</Heading>`.
> 
> The only way I'm able to get rid of type errors, is to declare _only_ a dummy `target={new Element}` prop. Introducing any other props results in the mentioned type errors.
> 
> I'm guessing the problem goes beyond the special `target` prop and the type checking seems to be more generally broken at the moment - or am I missing something?

I've concluded, that my problem is unrelated to this issue. Using components from "carbon-components-svelte" is working as expected. It seems to be a problem with my bunlding configuration, which is odd because it's a fresh SvelteKit library project.

**Comment by silveltman** [comment]
@pfloride I'm actually running into the same issue. I'm creating a component library in sveltekit, so a 'frech Sveltekit Libary' too.

Do you have any progress? If you do, please let me know :)

**Comment by pfloride** [comment]
@silveltman Sorry for the late reply. I solved the problem by ditching SvelteKit altogether and adopting a rollup based setup based on what's been done in https://github.com/carbon-design-system/carbon-components-svelte. Works like a charm, I can recommend!

@Princesseuh Feel free to hide these comments if they're too off-topic.

**Comment by silveltman** [comment]
@pfloride Hmm, interesting. I use [Histoire.dev](https://histoire.dev/) for creating stories, which is Vite based. Would you still recommend it then?

**Comment by pfloride** [comment]
@silveltman I use https://storybook.js.org/, there I could just split the library from the storybook installation, then the two build-systems (library/storybook) don't have to match at all, as long as both can handle the required features of your lib components. That's probably also possible with histoire. Just try to move all the histoire code (and the package.json) into a standalone folder and import the library components from the root folder.

**Comment by silveltman** [comment]
@pfloride Is there a reason you chose a rollup based setup and not Vite? 

Btw, I'm looking into turborepo, so I can have a Histoire app and the svelte library running side-by-side

**Comment by pfloride** [comment]
@silveltman No, not really, I just needed a working setup that ticks all the boxes.

**Comment by andywwright** [comment]
Just got the very same issue with my own svelte component inside astro project and it seems that just restarting vscode has helped.

Not sure if it is related, but at the same time I had another mind boggling issue - `import` statement of just created and perfectly working `Layout` was underlined in red and saying

`Cannot find module "../layouts/Layout.astro" or its corresponding type declarations.ts(2307)`

which went away after vscode reloading as well.

Immediately I've created a few fresh astro-svelte projects, but for the life of me I couldn't reproduce the issue anymore.

**Comment by silveltman** [comment]
@andywwright the problem with importing svelte components into my .Astro files too to come and go in my case. 

As for the second error you mentioned, probably a TS error. Ts seems to only update the type declarations on restart. Most definitely there is a solution for this, but I haven't looked into it. 

What project structure do you use is term of, where is you app vs your library located and how do they interact?

**Comment by andywwright** [comment]
it was literally an empty test project. all I have done by that time was
```
pnpm create astro astro-svelte-test

// Do you plan to write TypeScript? - No

cd astro-svelte-test
pnpm astro add svelte

// in src folder
mkdir layouts
new file Layout.astro
mkdir islands
new file Form.svelte

// index astro
---
import Layout from '../layouts/Layout.astro'
import Form from '../islands/Form.svelte'
---

<Layout>
	<Form client:load />
</Layout>
```
bum! errors about target in form and having no idea what Layout is

**Comment by Princesseuh** [comment]
The Astro error is a regression in the latest version, I'm still investigating where it could come from, as I've not been able to reproduce it myself

**Comment by MarcusOtter** [comment]
Running into this issue on the latest version with [`svelte-flatpickr`](https://github.com/jacobmischka/svelte-flatpickr) and I can't seem to get the workaround to work (guess the component is not exported).

My error is always
```
Cannot find module 'svelte-flatpickr' or its corresponding type declarations. ts(2307)
```
and it does not go away when I restart VSCode

**Is there anything we can contribute with to the packages themselves to make them more compatible with Astro?**

Because I had no issues using the `DatePicker` from [`carbon-components-svelte`](https://github.com/carbon-design-system/carbon-components-svelte) even though that also uses [`flatpickr`](https://github.com/flatpickr/flatpickr) under the hood (just like [`svelte-flatpickr`](https://github.com/jacobmischka/svelte-flatpickr)). So there is some difference in the libs that makes one of them work without hassle, but the other one not. They both seem to be using `SvelteComponentTyped`.

**Comment by Princesseuh** [comment]
That package is very weirdly packaged, it only has a single export using a non-standard export name and it declares its types using a fake module

https://github.com/jacobmischka/svelte-flatpickr/blob/master/package.json#L32-L36

I don't really know how we would support that without implementing Svelte's module resolution I assume. If the package just exposed the files normally, you could just import the Svelte components.

**Comment by vonagam** [comment]
For me shown workaround wasn't working. I was importing `flowbite-svelte` like this:
```ts
import Button from "flowbite-svelte/Button.svelte";
```
I guess it must be because the package uses `exports` and provides "types" for each one, so it is not a raw/direct access which I assume the workaround is about.

Anyway, just to add another remedy for those who stumble here: as mentioned in the issue a potential solution is to patch the definition, so I've simply added such patch to `src/env.d.ts` with snippet like this:
```ts
declare module 'svelte' {
  export interface ComponentConstructorOptions {
    target?: Element | Document | ShadowRoot;
  }
}
```
